Fix "None" map name retrieved for generated maps

Katharsas opened the pull request:
Fixes #1772

Katharsas committed:
Fix "None" map name retrieved for generated maps

1-alex98 commented on src/main/java/com/faforever/client/replay/ReplayService.java:
```diff
@@ -231,6 +233,7 @@ static Integer parseSupComVersion(byte[] rawReplayBytes) {
   static String parseMapName(byte[] rawReplayBytes) {
     int mapDelimiterIndex = Bytes.indexOf(rawReplayBytes, new byte[]{0x00, 0x0D, 0x0A, 0x1A});
     String mapPath = new String(rawReplayBytes, MAP_NAME_OFFSET, mapDelimiterIndex - MAP_NAME_OFFSET, US_ASCII);
+    // TODO make sure that these are valid characters for a file name? this is potentially unsafe user input
```
Can u specify what you mean, maybe link a GitHub issue. Donot think that a user could use this string to attack another client

Katharsas replied:
The usual maps dont require calling this beause mapName is in metadata. That mapName is retrived from some server column by matching via filename. So for normal maps we are guarantueed to get server entries. When directly parsing, we are reading a name that comes directly from one players client.

I guess it doesnt matter because downloading the mapo will simply fail if we have invalid characters right?

Sheikah45 replied:
I added in a check for invalid filename characters

1-alex98 commented on src/main/java/com/faforever/client/replay/ReplayService.java:
```diff
@@ -503,6 +507,14 @@ private void runFafReplayFile(Path path) throws IOException {
 
+    // For map generator games the map name is "None" because replay server gets map name by from DB based on filename
+    // from replay data, and DB does not contain generated maps.
+    if (StringUtils.equalsIgnoreCase(mapName, "None")) {
```
This is also the case for other featured mods like coop.  How would that work out there?

Katharsas replied:
Coop maps seem to be handled in the lines directly above this code. Maybe it was null or empty in the past and due to a replay server change is "None" now for coop maps as well?

1-alex98 commented on src/main/java/com/faforever/client/replay/ReplayService.java:
```diff
@@ -503,6 +507,14 @@ private void runFafReplayFile(Path path) throws IOException {
       mapName = parseMapFolderName(rawReplayBytes);
     }
 
+    // For map generator games the map name is "None" because replay server gets map name by from DB based on filename
+    // from replay data, and DB does not contain generated maps.
+    if (StringUtils.equalsIgnoreCase(mapName, "None")) {
+      String maybeMapGen = parseMapFolderName(rawReplayBytes);
+      if (mapGeneratorService.isGeneratedMap(maybeMapGen)) {
+        mapName = maybeMapGen;
+      }
+    }
```
Could we write a unit test for that?

Katharsas replied:
Do we have some actual specification that determines when mapName is supposed to be null, empty or "None"? Otherwise the test will just enforce arbitrary things.

Katharsas replied:
Sheikah added a test

1-alex98 commented:
CI failing

Sheikah45 committed:
Add check for invalid characters in filename

Sheikah45 committed:
Add unit tests for invalid filenames

Sheikah45 committed:
Add unit test for generated map check

Katharsas commented:
@Sheikah45 thanks for improving this, looking good imo.

Sheikah45 committed:
Use parseMapName over parseMapFolderName

For some reason parseMapFolderName returns the map name without any capitalization from the raw replay bytes. This causes issues with the base64 string encoding used by the next gen map generator. parseMapName returns the map name with correct capitalization making it more suitable for usage. As an aside it appears that parseMapFolderName uses the scenario file in the raw bytes to get the map name whereas parseMapName uses the .scmap file and the capitalization differences are reflected there.

Sheikah45 committed:
Update unit tests
